src/DotNetProjectFile.Analyzers/Analyzers/MsBuild/ThirdPartyLicenseResolver.cs

```diff
@@ -7,32 +6,56 @@ namespace DotNetProjectFile.Analyzers.MsBuild;
 [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
 public sealed class ThirdPartyLicenseResolver() : MsBuildProjectFileAnalyzer(
     Rule.OnlyIncludePackagesWithExplicitLicense,
-    Rule.PackageOnlyContainsDeprecatedLicenseUrl)
+    Rule.PackageOnlyContainsDeprecatedLicenseUrl,
+    Rule.PackageIncompatibleWithProjectLicense)
 {
     /// <inheritdoc />
     public override bool DisableOnFailingImport => false;
 
     /// <inheritdoc />
-    protected override void Register(ProjectFileAnalysisContext<MsBuildProject> context)
+    protected override void Register(ProjectFileAnalysisContext context)
     {
+        var projectLicense = Licenses.FromExpression(context.GetMsBuildProperty("PackageLicenseExpression"));
+
         foreach (var reference in context.File.ItemGroups.SelectMany(g => g.Children)
             .OfType<PackageReferenceBase>()
             .Where(r => r.Version is { Length: > 0 }))
         {
-            if (reference.GetLicensedPackage() is not { } package)
-            {
-                context.ReportDiagnostic(Rule.OnlyIncludePackagesWithExplicitLicense, reference, reference.IncludeOrUpdate);
-            }
-            else if (package.UrlOnly() && Licenses.FromUrl(package.LicenseUrl) == Licenses.Unknown)
-            {
-                context.ReportDiagnostic(Rule.PackageOnlyContainsDeprecatedLicenseUrl, reference, reference.IncludeOrUpdate);
-            }
+            Report(reference, projectLicense, context);
+        }
+    }
+
+    private static void Report(PackageReferenceBase reference, LicenseExpression projectLicense, ProjectFileAnalysisContext context)
+    {
+        if (reference.GetLicensedPackage() is not { } package)
+        {
+            context.ReportDiagnostic(Rule.OnlyIncludePackagesWithExplicitLicense, reference, reference.IncludeOrUpdate);
+            return;
+        }
+
+        var packageLicense = package.LicenseExpression();
+
+        if (package.UrlOnly() && packageLicense.IsUnknown)
+        {
+            context.ReportDiagnostic(Rule.PackageOnlyContainsDeprecatedLicenseUrl, reference, reference.IncludeOrUpdate);
+        }
+        else if (projectLicense.IsKnown && !packageLicense.CompatibleWith(projectLicense))
```

Comment:
We shouldn't need this `projectLicense.IsKnown` here, I would argue it should be the responsibility of the different licenses to determine if they can be used in any case in their `CompatibleWith` implementation.

Reply:
That is true once we have a good way to deal with:
- LicenseExpression None is supported
- CustomLicense (LicenseFile is specified which should be compatible then)

I would like to do that in follow-up PR's.

Reply:
I fail to see how that is relevant. Permissive ones will always return `true` and the strict copyleft ones will return `false`, when the license is `None`, `Custom` or `Unknown`, so that's not different from the current behaviour of `Unknown`

Reply:
Well, with `None` it should be true for all, as you do not share anything. But we can say that we will not ship in the meantime, so that is not an issue.

Reply:
I would argue against that, because if a dependency is GPL3 and your own license is None, you are in violation of GPL3